subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -426,6 +430,33 @@ private void configureConfigurations(Project project, JavaPluginExtension extens
 
         apiElementsConfiguration.deprecateForDeclaration(IMPLEMENTATION_CONFIGURATION_NAME, COMPILE_ONLY_CONFIGURATION_NAME);
         runtimeElementsConfiguration.deprecateForDeclaration(IMPLEMENTATION_CONFIGURATION_NAME, COMPILE_ONLY_CONFIGURATION_NAME, RUNTIME_ONLY_CONFIGURATION_NAME);
+
+        createSourcesVariant(project, extension);
+    }
+
+    private Configuration createSourcesVariant(Project project, JavaPluginExtension java) {
+        SourceSet main = java.getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+
+        final Configuration variant = project.getConfigurations().create(SOURCE_ELEMENTS_VARIANT_NAME);
+        variant.setVisible(false);
+        variant.setCanBeResolved(false);
+        variant.setCanBeConsumed(true);
+
+        final ObjectFactory objects = project.getObjects();
+        variant.attributes(attributes -> {
+            attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.VERIFICATION));
+            attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.SOURCES));
+            attributes.attribute(Sources.SOURCES_ATTRIBUTE, objects.named(Sources.class, Sources.ALL_SOURCE_DIRS));
+        });
+
+        // TODO: 18791 - This is a hack until we have a proper way of injecting ObjectFactory
+        ((DefaultConfigurationPublications) variant.getOutgoing()).setObjectFactory(objects);
+
+        variant.getOutgoing().artifacts(main.getAllSource().getSourceDirectories().getElements().flatMap(e -> project.provider(() -> e)), artifact -> {
```

Comment:
What's the extra levels of indirection here for?  getElements returns a provider, flatMap returns a provider and project.provider() creates another new provider.  Can we not just pass `getElements()` here?